nvflare/fuel/f3/streaming/byte_streamer.py

```diff
@@ -71,17 +82,162 @@ def __init__(
+                StreamHeaderKey.SEQUENCE: self.seq,
+                StreamHeaderKey.OFFSET: self.offset,
+                StreamHeaderKey.OPTIONAL: self.optional,
+            }
+        )
+
+        errors = self.cell.fire_and_forget(
+            STREAM_CHANNEL, STREAM_DATA_TOPIC, self.target, message, secure=self.secure, optional=self.optional
+        )
+        error = errors.get(self.target)
+        if error:
+            msg = f"{self} Message sending error to target {self.target}: {error}"
+            log.debug(msg)
+            self.stop(StreamError(msg))
+            return
+
+        # Update state
+        self.seq += 1
+        self.offset += self.buffer_size
+        self.buffer_size = 0
+        self.direct_buf = None
+
+        # Update future
+        self.stream_future.set_progress(self.offset)
+
+    def stop(self, error: Optional[StreamError] = None, notify=True):
+
+        self.stopped = True
+
+        if not error:
+            # Result is the number of bytes streamed
+            if self.stream_future:
+                self.stream_future.set_result(self.offset)
+            return
+
+        # Error handling
+        log.debug(f"{self} Stream error: {error}")
+        if self.stream_future:
+            self.stream_future.set_exception(error)
+
+        if notify:
+            message = Message(None, None)
+
+            if self.headers:
+                message.add_headers(self.headers)
+
+            message.add_headers(
+                {
+                    StreamHeaderKey.STREAM_ID: self.sid,
+                    StreamHeaderKey.DATA_TYPE: StreamDataType.ERROR,
+                    StreamHeaderKey.OFFSET: self.offset,
+                    StreamHeaderKey.ERROR_MSG: str(error),
+                }
+            )
+            self.cell.fire_and_forget(
+                STREAM_CHANNEL, STREAM_DATA_TOPIC, self.target, message, secure=self.secure, optional=True
+            )
+
+    def handle_ack(self, message: Message):
+
+        origin = message.get_header(MessageHeaderKey.ORIGIN)
+        offset = message.get_header(StreamHeaderKey.OFFSET, None)
+        error = message.get_header(StreamHeaderKey.ERROR_MSG, None)
+
+        if error:
+            self.stop(StreamError(f"{self} Received error from {origin}: {error}"), notify=False)
+            return
+
+        if offset > self.offset_ack:
+            self.offset_ack = offset
+
+        if not self.ack_waiter.is_set():
+            self.ack_waiter.set()
+
 
 class ByteStreamer:
     def __init__(self, cell: CoreCell):
         self.cell = cell
         self.cell.register_request_cb(channel=STREAM_CHANNEL, topic=STREAM_ACK_TOPIC, cb=self._ack_handler)
         self.tx_task_map = {}
         self.map_lock = threading.Lock()
+        self.chunk_size = CommConfigurator().get_streaming_chunk_size(STREAM_CHUNK_SIZE)
 
         self.sent_stream_counter_pool = StatsPoolManager.add_counter_pool(
```

Comment:
Oh No! You are creating one pool for each ByteStreamer object. There will be many pool objects created and wasted. You should create one such pool for the whole process. Otherwise, the results in the pool won't be correct.

Reply:
Changed the pool to be static and moved cell name from scope to category.